File: regular_expression.rb
$LOAD_PATH << '.'
require 'DFA&NFA.rb'

module Pattern
    def bracket(outer_precedence)
        if self.precedence < outer_precedence
            '(' + self.to_s + ')'
        else
            self.to_s
        end
    end

    def inspect
        "/#{self}/"
    end

    def matches?(string)
        self.to_nfa_design.accepts?(string)
    end
end

class Empty
    include Pattern

    def to_s
        ''
    end

    def precedence
        3
    end

    def to_nfa_design
        start_state = Object.new
        accept_states = [start_state]
        rulebook = NFARulebook.new([])

        NFADesign.new(start_state, accept_states, rulebook)
    end
end

class Literal < Struct.new(:character)
    include Pattern

    def to_s
        self.character
    end

    def precedence
        3
    end

    def to_nfa_design
        start_state = Object.new
        accept_state = Object.new
        rule = FARule.new(start_state, character, accept_state)
        rulebook = NFARulebook.new([rule])

        NFADesign.new(start_state, [accept_state], rulebook)
    end
end

class Concatenate < Struct.new(:first, :second)
    include Pattern

    def to_s
        [self.first, self.second].map { |pattern| pattern.bracket(self.precedence) }.join
    end

    def precedence
        1
    end

    def to_nfa_design
        first_nfa_design = self.first.to_nfa_design
        second_nfa_design = self.second.to_nfa_design

        start_state = first_nfa_design.start_state
        accept_states = second_nfa_design.accept_states

        rules = first_nfa_design.rulebook.rules + second_nfa_design.rulebook.rules
        extra_rules = first_nfa_design.accept_states.map { |state|
            FARule.new(state, nil, second_nfa_design.start_state)
        }

        rulebook = NFARulebook.new(rules + extra_rules)
        
        NFADesign.new(start_state, accept_states, rulebook)
    end
end

class Choose < Struct.new(:first, :second)
    include Pattern

    def to_s
        [self.first, self.second].map { |pattern| pattern.bracket(self.precedence) }.join('|')
    end

    def precedence
        0
    end

    def to_nfa_design
        first_nfa_design = self.first.to_nfa_design
        second_nfa_design = self.second.to_nfa_design

        start_state = Object.new
        accept_states = first_nfa_design.accept_states + second_nfa_design.accept_states

        rules = first_nfa_design.rulebook.rules + second_nfa_design.rulebook.rules
        extra_rules = [first_nfa_design, second_nfa_design].map { |nfa_design| 
            FARule.new(start_state, nil, nfa_design.start_state)
        }

        rulebook = NFARulebook.new(rules + extra_rules)

        NFADesign.new(start_state, accept_states, rulebook)
    end
end

class Repeat < Struct.new(:pattern)
    include Pattern

    def to_s
        self.pattern.bracket(self.precedence) + '*'
    end

    def precedence
        2
    end

    def to_nfa_design
        pattern_nfa_design = pattern.to_nfa_design

        start_state = Object.new
        accept_states = pattern_nfa_design.accept_states + [start_state]
        rules = pattern_nfa_design.rulebook.rules
        extra_rules = pattern_nfa_design.accept_states.map { |accept_state| 
            FARule.new(accept_state, nil, pattern_nfa_design.start_state)
        } + [FARule.new(start_state, nil, pattern_nfa_design.start_state)]
        rulebook = NFARulebook.new(rules + extra_rules)

        NFADesign.new(start_state, accept_states, rulebook)
    end
end

pattern = Repeat.new(
    Concatenate.new(
        Literal.new('a'),
        Choose.new(Empty.new, Literal.new('b'))
    )
)

puts pattern.matches?('ababaaa')

File: regular_expression.mjs
import {
    FARule,
    NFARulebook,
    NFADesign
} from './DFA&NFA.mjs'

class Pattern {
    bracket(outer_precedence) {
        if (this.precedence < outer_precedence) {
            return `(${this.to_s()})`;
        } else {
            return this.to_s();
        }
    }

    matches(string) {
        // to_nfa_design() should be a recursion method for each RE pattern class
        return this.to_nfa_design().accepts(string);
    }
}

class Empty extends Pattern {
    constructor() {
        super();
        this.to_s = () => '';
        this.precedence = 3;
    }

    to_nfa_design() {
        let start_state = new Object;
        let accept_states = [start_state];
        let rulebook = new NFARulebook([]);

        return new NFADesign(start_state, accept_states, rulebook);
    }
}

class Literal extends Pattern {
    constructor(character) {
        super();
        this.character = character;
        this.to_s = () => this.character;
        this.precedence = 1;
    }

    to_nfa_design() {
        // let start_state = Math.random(); // from a rough point of view, any two random numbers are unequal to each others
        let start_state = new Object(); // but any two new Object()s are absolutely unequal to each others

        // accept_state should be an array
        // let accept_state = [ Math.random() ];
        let accept_state = [new Object()];
        let rule = new FARule(start_state, this.character, accept_state);
        let rulebook = new NFARulebook([rule]);
        return new NFADesign(start_state, accept_state, rulebook);
    }
}

class Concatenate extends Pattern {
    constructor(first, second) {
        super();
        this.first = first;
        this.second = second;
        this.to_s = () => {
            let arr = [this.first, this.second].map(ele => ele.bracket(this.precedence));
            return arr.join('');
        };
        this.precedence = 1;
    }

    to_nfa_design() {
        let first_nfa_design = this.first.to_nfa_design();
        let second_nfa_design = this.second.to_nfa_design();

        let start_state = first_nfa_design.start_state;
        let accept_states = second_nfa_design.accept_states;

        let rules = [...first_nfa_design.rulebook.rules, ...second_nfa_design.rulebook.rules]
        let extra_rules = first_nfa_design.accept_states.map(state => {
            return new FARule(state, null, second_nfa_design.start_state);
        });
        let rulebook = new NFARulebook([...rules, ...extra_rules]);

        return new NFADesign(start_state, accept_states, rulebook);
    }
}

class Choose extends Pattern {
    constructor(first, second) {
        super();
        this.first = first;
        this.second = second;
        this.to_s = () => {
            let arr = [this.first, this.second].map(ele => ele.bracket(this.precedence));
            return arr.join('|');
        }
        this.precedence = 0;
    }

    to_nfa_design() {
        let first_nfa_design = this.first.to_nfa_design();
        let second_nfa_design = this.second.to_nfa_design();

        let start_state = new Object();
        // accept_state should be an array
        // let accept_states = [ new Object() ];
        let accept_states = [...first_nfa_design.accept_states, ...second_nfa_design.accept_states];

        let rules = [...first_nfa_design.rulebook.rules, ...second_nfa_design.rulebook.rules]
        let extra_rules1 = [first_nfa_design, second_nfa_design].map(nfa_design => {
            return new FARule(start_state, null, nfa_design.start_state);
        });

        // let extra_rules2 = [...first_nfa_design.accept_states, ...second_nfa_design.accept_states].map(state => {
        //     return new FARule(state, null , accept_state[0]);
        // });

        let rulebook = new NFARulebook([...rules, ...extra_rules1]);
        return new NFADesign(start_state, accept_states, rulebook);
    }
}

class Repeat extends Pattern {
    constructor(pattern) {
        super();
        this.pattern = pattern;
        this.to_s = () => this.pattern.bracket(this.precedence) + '*';
        this.precedence = 2;
    }

    to_nfa_design() {
        let pattern_nfa_design = this.pattern.to_nfa_design();
        let start_state = new Object();
        let accept_states = [...pattern_nfa_design.accept_states, ...[start_state]];

        let rules = pattern_nfa_design.rulebook.rules;
        let extra_rules1 = pattern_nfa_design.accept_states.map(accept_state => {
            return new FARule(accept_state, null, start_state);
        });
        let extra_rules2 = [new FARule(start_state, null, pattern_nfa_design.start_state)];

        let rulebook = new NFARulebook([...rules, ...extra_rules1, ...extra_rules2]);
        return new NFADesign(start_state, accept_states, rulebook);
    }
}

let pattern = new Repeat(
    new Concatenate(
        new Literal('a'),
        new Choose(new Empty(), new Literal('b'))
    )
);
console.log(pattern.to_s());
console.log(pattern.matches('abaaaaaaab'));